qiskit_experiments/curve_analysis/grouped_curve_analysis.py

```diff
@@ -0,0 +1,324 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Analysis class for multi-group curve fitting.
+"""
+# pylint: disable=invalid-name
+
+from typing import Dict, List, Tuple, Optional, Any
+from collections import defaultdict
+
+import lmfit
+import numpy as np
+from uncertainties import unumpy as unp, UFloat
+
+from qiskit_experiments.framework import ExperimentData, AnalysisResultData, Options
+from qiskit_experiments.exceptions import AnalysisError
+
+from .base_curve_analysis import BaseCurveAnalysis, PARAMS_ENTRY_PREFIX
+from .curve_data import CurveFitResult
+from .utils import analysis_result_to_repr, eval_with_uncertainties
+
+
+class MultiGroupCurveAnalysis(BaseCurveAnalysis):
+    r"""Curve analysis with multiple independent groups.
+
+    The :class:`.MultiGroupCurveAnalysis` provides a fitting framework for the case
+
+    .. math::
+
+        \Theta_{i, \mbox{opt}} = \arg\min_{\Theta_{i, \mbox{fit}}} (F(X_i, \Theta_i)-Y_i)^2,
+
+    where :math:`F` is a common fit model for multiple independent dataset :math:`(X_i, Y_i)`.
+    The fit generates multiple fit outcomes :math:`\Theta_{i, \mbox{opt}}` for each dataset.
+
+    Each experiment circuit must have metadata of data sorting keys for model mapping and
+    group mapping. For example, if the experiment has two experiment conditions "A" and "B"
+    while fitting the outcomes with three models "X", "Y", "Z", then the metadata must consist
+    of the condition and model, e.g. ``{"condition": "A", "model": "X"}``.
+
+    In this example, "condition" should be defined in the ``group_data_sort_key``
+    of the :class:`.MultiGroupCurveAnalysis`, and "model" should be defined in
+    the ``data_sort_key`` for each LMFIT model.
+
+    .. code-block:: python
+
+        import lmfit
+        from qiskit_experiments.curve_analysis import MultiGroupCurveAnalysis
+
+        analysis = MultiGroupCurveAnalysis(
+            groups=["groupA", "groupB"],
```

Comment:
How does group relate to the metadata in the circuits?

Reply:
They are classified by `group_data_sort_key `, i.e. `groups[0]` must use `group_data_sort_key[0]` to filter circuits based on the metadata.